man/en/sos.1
.TH sos 1 "April 2020"

.SH NAME
sos \- A unified tool for collecting system logs and other debug information
.SH SYNOPSIS
\fBsos\fR component [options]

.SH DESCRIPTION
\fBsos\fR is a diagnostic data collection utility, used by system administrators,
support representatives, and the like to assist in troubleshooting issues with
a system or group of systems.

The most well known function is \fBsos report\fR.

An sos archive is typically requested by support organizations to collect baseline
configuration and system data from which to begin the troubleshooting process.


.SH COMPONENTS

sos supports several subcommands or components. Each provides a different set
of information for the user. Supported components are as follows

.TP
.B report
Report generates an archive of system information including configuration files
and command output. Information included in the report is based upon plugins that
are activated automatically when certain criteria, such as installed packages, files,
services, or system architecture is detected.

See \fBsos report --help\fR and \fBman sos-report\fR for more information.

May also be invoked via the alias \fBrep\fR.

.TP
.B collect
Collect is used to capture reports on multiple systems simultaneously. These
systems can either be defined by the user at the command line and/or defined by
clustering software that exists either on the local system or on a "primary" system
that is able to inform about other nodes in the cluster.

When running collect, sos report will be run on the remote nodes, and then the
resulting archives will be copied from those nodes to the local system running
sos collect. Archives are then removed from the remote systems.

See \fBsos collect --help\fR and \fBman sos-collect\fR for more information.

May also be invoked via the alias \fBsos collector\fR or the deprecated command
\fBsos-collector\fR.

.TP
.B clean|cleaner|mask
This subcommand takes input of either 1) an sos report tarball, 2) a collection
of sos report tarballs such as from \fBcollect\fR, or 3) the unpackaged
directory of an sos report and obfuscates potentially sensitive system information
that is not covered by the standard postprocessing of \fBsos report\fR.

Such data includes IP addresses, networks, MAC addresses, and more. Data obfuscated
by this command will remain consistent throughout the report and across reports provided
in the same invocation. Additionally, care is taken to maintain network topology relationships
between matched data items.

See \fB sos clean --help\fR and \fBman sos-clean\fR for more information.

May be invoked via either \fBsos clean\fR, \fBsos cleaner\fR, \fBsos mask\fR,
or via the \fB--clean\fR, \fB--cleaner\fR or \fB --mask\fR options
for \fBreport\fR and \fBcollect\fR.

.TP
.B help
This subcommand is used to retrieve more detailed information on the various SoS
commands and components than is directly available in either other manpages or
--help output.

See \fB sos help --help\fR and \fB man sos-help\fR for more information.

.TP
.B upload
This subcommand uploads an input file to either a distribution-defined or a
user-defined target. The file can be an SOS archive or any other type of file,
such as a vmcore or log file.

See \fB sos upload --help\fR and \fB man sos-upload\fR for more information.

.SH GLOBAL OPTIONS
sos components provide their own set of options, however the following are available
to be set across all components.

.B \-\-batch
Do not prompt interactively, user will not be prompted for any data
.TP
.B \-\-encrypt
Encrypt the resulting archive, and determine the method by which that encryption
is done by either a user prompt or environment variables.

When run with \fB--batch\fR, using this option will cause sos to look for either the
\fBSOSENCRYPTKEY\fR or \fBSOSENCRYPTPASS\fR environment variables. If set, this will
implicitly enable the \fB--encrypt-key\fR or \fB--encrypt-pass\fR options, respectively,
to the values set by the environment variable. This enables the use of these options
without directly setting those options in a config file or command line string. Note that
use of an encryption key has precedence over a passphrase.

Otherwise, using this option will cause sos to prompt the user to choose the method
of encryption to use. Choices will be [P]assphrase, [K]ey, [E]nv vars, or [N]o encryption.
If passphrase or key the user will then be prompted for the respective value, env vars will
cause sos to source the information in the manner stated above, and choosing no encryption
will disable encryption.

See the sections on \fB--encrypt-key\fR and \fB--encrypt-pass\fR below for more
information.
.TP
.B \--encrypt-key KEY
Encrypts the resulting archive that sos report produces using GPG. KEY must be
an existing key in the user's keyring as GPG does not allow for keyfiles.
KEY can be any value accepted by gpg's 'recipient' option.

Note that the user running sos report must match the user owning the keyring
from which keys will be obtained. In particular this means that if sudo is
used to run sos report, the keyring must also be set up using sudo
(or direct shell access to the account).

Users should be aware that encrypting the final archive will result in sos
using double the amount of temporary disk space - the encrypted archive must be
written as a separate, rather than replacement, file within the temp directory
that sos writes the archive to. However, since the encrypted archive will be
the same size as the original archive, there is no additional space consumption
once the temporary directory is removed at the end of execution.

This means that only the encrypted archive is present on disk after sos
finishes running.

If encryption fails for any reason, the original unencrypted archive is
preserved instead.
.TP
.B \--encrypt-pass PASS
The same as \--encrypt-key, but use the provided PASS for symmetric encryption
rather than key-pair encryption.
.TP
.B \--config-file CONFIG
Specify alternate configuration file.
.TP
.B \-s, \--sysroot SYSROOT
Specify an alternate root file system path.
.TP
.B \--tmp-dir DIRECTORY
Specify alternate temporary directory to copy data during execution.
.TP
.B \--threads THREADS
Specify the number of threads sos report will use for concurrency. Defaults to 4. 
.TP
.B \-v, \--verbose
Increase logging verbosity. May be specified multiple times to enable
additional debugging messages.

The following table summarizes the effects of different verbosity levels:

    1 (-v)   :  Enable debug messages for sos.log. Show individual plugins starting.
    2 (-vv)  :  Also print debug messages to console.
    3 (-vvv) :  Enable debug messages for archive file operations. Note this will dramatically
                increase the amount of logging.

.TP
.B \-q, \--quiet
Only log fatal errors to stderr.
.TP
.B \-z, \-\-compression-type {auto|xz|gzip|none}
Compression type to use when compressing the final archive output. If the value \fBnone\fP is given
the tar archive will be written directly with no compression.
.TP
.B \--help
Display usage message.
.SH SEE ALSO
.BR sos.conf (5)
.SH MAINTAINER
.nf
Maintained on GitHub at https://github.com/sosreport/sos
.fi
.SH AUTHORS & CONTRIBUTORS
See \fBAUTHORS\fR file in the package documentation.


src/ChangeLog
2007-06-15  Eugene Teo  <eteo@redhat.com>

	* lib/sos/plugins/apache.py, lib/sos/plugins/nfsserver.py, lib/sos/plugins/selinux.py, lib/sos/plugins/xinetd.py, lib/sos/plugins/ssh.py, lib/sos/plugins/sendmail.py, lib/sos/plugins/samba.py, lib/sos/plugins/named.py, lib/sos/plugins/cluster.py:
	- Edited apache.py to gather /var/log/httpd/* log files
	- Added nfsserver.py to gather NFS server-related debugging information
	- Edited selinux.py to gather /etc/selinux/* configuration files
	- Added xinetd.py to gather xinetd-related information
	- Added ssh.py to gather ssh-related information
	- Added sendmail.py to gather sendmail information
	- Edited samba.py to gather /var/log/samba/* log files
	- Edited named.py to gather /etc/sysconfig/named
	- Edited cluster.py to gather the output of fdisk -l to show the
	shared storage devices that should be available to each system

2007-05-28  Eugene Teo  <eteo@redhat.com>

	* lib/sos/plugins/systemtap.py:
	- Added systemtap.py to gather SystemTap pre-requisites information

2007-05-28  Eugene Teo  <eteo@redhat.com>

	* lib/sos/plugins/amd.py:
	- Added amd.py to gather Amd automounter information

2007-05-25  Eugene Teo  <eteo@redhat.com>

	* lib/sos/plugins/xen.py, lib/sos/plugins/pam.py, lib/sos/plugins/memory.py:
	- Edited xen.py to determine if CPU has PAE/Intel VT/AMD-V support
	- Edited pam.py to gather configurations in /etc/security, and files
	  listing of /lib/security/pam_*so
	- Edited memory.py to gather /proc/{vmstat,slabinfo}, and free -m
	  output

2007-04-23  Navid Sheikhol-Eslami <navid@redhat.com>

	* Running "multipath" without arguments might change the device-mapper maps, which we want to avoid if things are broken.

2007-04-02  Navid Sheikhol-Eslami <navid@redhat.com>

	* Replaced xen plugin with (better) version from Chris Lalancette <clalance@redhat.com>

2007-03-29  Navid Sheikhol-Eslami <navid@redhat.com>

	* Added a checkenabled() function which can be used to disable a plugin at run-time.
	* Disable the progress-bar if verbosity is enabled.

2007-03-27  Navid Sheikhol-Eslami <navid@redhat.com>

	* Fixed hardware plugin to use modules.pcimap instead of deprecated pcitable.
	* Added a random suffix to sosreport tree to avoid overwriting an existing tree with same name.
	* Better logging using python's logging module.
	* Verbose logs included in sosreport (sos_logs/sos.log)

2007-03-15  Navid Sheikhol-Eslami <navid@redhat.com>

	* Implemented a progress bar (RFE BZ#219672) which can be disabled from the command line.
	* Added check to see if the loaded module matches the copy on the file-system

2007-03-14  Navid Sheikhol-Eslami <navid@redhat.com>

	* fixed BZ#219877 (ncurses "Cancel" button makes sosreport exit)

2007-03-07  Navid Sheikhol-Eslami <navid@redhat.com>

	* Allow passing multiple comma-separated plugin names to -n (--noplugin) and -o (--onlyplugin) options.
	* Added further commands' output to gather from lvm_dump


2007-02-20  Navid Sheikhol-Eslami <navid@redhat.com>

	* Added a specialized plugin for device-mapper related configuration files and command outputs (device-mapper, LVM and multipath)
	* Added --onlyplugin option (-o) to selectively choose which plugins to load (complementary to existing --noplugin)
	* Exit if no valid plugin was selected (rather than building an empty sosreport).

2007-02-16  Navid Sheikhol-Eslami <navid@redhat.com>

	* Strip out the shared secret (bindpw) from /etc/ldap.conf
	* Collect parsed configuration tree directly from ccsd (useful for troubleshooting parsing issues)
	* Scamble password information for fencing devices.

2007-01-26  Navid Sheikhol-Eslami <navid@redhat.com>

	* Added doRegexSub() to be called in postproc() to apply a regexp substitution to files
	* Added radius plugin for freeradius data collection
	* Ask full name to prevent errors when moving the sos tree before packaging
	* Reformat tar file name is no ticket number is given

2006-06-19  Steve Conklin  <sconklin@tintin>

	* ChangeLog, LICENSE, setup.py, sos.spec:
	Added License file and bumped release

2006-06-08  dlehman  <dlehman@tintin>

	* example_plugins/example.py, example_plugins/fsusage.py, example_plugins/release.py, example_plugins/template.py, lib/sos/helpers.py, lib/sos/plugins/apache.py, lib/sos/plugins/bootloader.py, lib/sos/plugins/cluster.py, lib/sos/plugins/filesys.py, lib/sos/plugins/ftp.py, lib/sos/plugins/general.py, lib/sos/plugins/hardware.py, lib/sos/plugins/kernel.py, lib/sos/plugins/ldap.py, lib/sos/plugins/libraries.py, lib/sos/plugins/mail.py, lib/sos/plugins/memory.py, lib/sos/plugins/named.py, lib/sos/plugins/networking.py, lib/sos/plugins/pam.py, lib/sos/plugins/process.py, lib/sos/plugins/rhn.py, lib/sos/plugins/rpm.py, lib/sos/plugins/samba.py, lib/sos/plugins/selinux.py, lib/sos/plugins/squid.py, lib/sos/plugins/startup.py, lib/sos/plugins/system.py, lib/sos/plugins/tarball.py, lib/sos/plugins/x11.py, lib/sos/plugintools.py, lib/sos/policyredhat.py, setup.py, sosreport:
	- Flesh out rhn plugin to handle Proxy or Satellite
	- Add package queries to policyredhat.py (allPkgsByName, pkgByName, pkgNVRA)
	- Add policy instance to the commons dict for access from plugins
	- Use string objects' methods instead of the string module where possible
	- Remove imports of unused string module
	- Cleanup some typos, redundant initializations, &c

2006-06-08  dlehman  <dlehman@tintin>


sos/cleaner/archives/sos.py
# Copyright 2021 Red Hat, Inc. Jake Hunsaker <jhunsake@redhat.com>

# This file is part of the sos project: https://github.com/sosreport/sos
#
# This copyrighted material is made available to anyone wishing to use,
# modify, copy, or redistribute it subject to the terms and conditions of
# version 2 of the GNU General Public License.
#
# See the LICENSE file in the source distribution for further information.

import os
import tarfile

from sos.cleaner.archives import SoSObfuscationArchive


class SoSReportArchive(SoSObfuscationArchive):
    """This is the class representing an sos report, or in other words the
    type the archive the SoS project natively generates
    """

    type_name = 'report'
    description = 'sos report archive'

    @classmethod
    def check_is_type(cls, arc_path):
        try:
            return tarfile.is_tarfile(arc_path) and 'sosreport-' in arc_path
        except Exception:
            return False


class SoSReportDirectory(SoSReportArchive):
    """This is the archive class representing a build directory, or in other
    words what `sos report --clean` will end up using for in-line obfuscation
    """

    type_name = 'report_dir'
    description = 'sos report directory'

    @classmethod
    def check_is_type(cls, arc_path):
        if os.path.isdir(arc_path):
            return 'sos_logs' in os.listdir(arc_path)
        return False


class SoSCollectorArchive(SoSObfuscationArchive):
    """Archive class representing the tarball created by ``sos collect``. It
    will not provide prep files on its own, however it will provide a list
    of SoSReportArchive's which will then be used to prep the parsers
    """

    type_name = 'collect'
    description = 'sos collect tarball'
    is_nested = True

    @classmethod
    def check_is_type(cls, arc_path):
        try:
            return (tarfile.is_tarfile(arc_path) and 'sos-collect' in arc_path)
        except Exception:
            return False

    def get_nested_archives(self):
        self.extract(quiet=True)
        _path = self.extracted_path
        archives = []
        for fname in os.listdir(_path):
            arc_name = os.path.join(_path, fname)
            if 'sosreport-' in fname and tarfile.is_tarfile(arc_name):
                archives.append(SoSReportArchive(arc_name, self.tmpdir,
                                                 self.keep_binary_files,
                                                 self.treat_certificates))
        return archives


class SoSCollectorDirectory(SoSCollectorArchive):
    """The archive class representing the temp directory used by ``sos
    collect`` when ``--clean`` is used during runtime.
    """

    type_name = 'collect_dir'
    description = 'sos collect directory'

    @classmethod
    def check_is_type(cls, arc_path):
        if os.path.isdir(arc_path):
            for fname in os.listdir(arc_path):
                if 'sos-collector-' in fname:
                    return True
        return False

# vim: set et ts=4 sw=4 :
